core/client/fs/src/main/java/alluxio/client/block/policy/CapacityBaseRandomPolicy.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.policy;
+
+import alluxio.client.block.BlockWorkerInfo;
+import alluxio.client.block.policy.options.GetWorkerOptions;
+import alluxio.wire.WorkerNetAddress;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Random load block base on capacity of worker,
+ * used for cluster of workers with different capacity.
+ * Then larger capacity worker will receive more load request.
+ *
+ */
+@ThreadSafe
+public class CapacityBaseRandomPolicy implements BlockLocationPolicy {
+
+  /**
+   * used for binary search.
+   */
+  private static final Comparator<CapacityRangeWorkerInfo> SEARCH_COMPARATOR = (o1, o2) -> {
+    if (o1.mStartInclusive <= o2.mStartInclusive && o1.mEndExclusive >= o2.mEndExclusive) {
+      return 0;
+    }
+    if (o1.mEndExclusive <= o2.mStartInclusive) {
+      return -1;
+    }
+    return 1;
+  };
+
+  @Nullable
+  @Override
+  public WorkerNetAddress getWorker(GetWorkerOptions options) {
+    Iterable<BlockWorkerInfo> blockWorkerInfos = options.getBlockWorkerInfos();
+    List<CapacityRangeWorkerInfo> workerList = new ArrayList<>();
+    long totalCapacity = 0L;
+    for (Iterator<BlockWorkerInfo> iterator = blockWorkerInfos.iterator(); iterator.hasNext(); ) {
+      BlockWorkerInfo workerInfo = iterator.next();
+      long currentCapacity = workerInfo.getCapacityBytes();
+      if (currentCapacity > 0) {
+        workerList.add(new CapacityRangeWorkerInfo(totalCapacity, totalCapacity + currentCapacity,
+            workerInfo));
+      }
+      totalCapacity += currentCapacity;
+    }
+    if (totalCapacity == 0L) {
+      return null;
+    }
+    long randomLong = randomInCapacity(totalCapacity);
+    //workerList already in ascending order, use binarySearch
+    int matchIndex = Collections
+        .binarySearch(workerList, new CapacityRangeWorkerInfo(randomLong, randomLong + 1, null),
```

Comment:
I might use a single value treemap to implement this algorithm rather than a range.  You could use the current capacity as the key and the worker index as the value.  Then you can use the lowerEntry(randomLong) to get the worker you want.

Reply:
Thanks for your good suggestion. I have done with much simplified implement.